Add volume migration crd cleanup go routine

chethanv28 opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR is adding a cleanup method in volumeMigrationService to handle deletion of stale crd instances.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #441 

**Special notes for your reviewer**:
This is a follow-up PR from the discussion here: 
https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/426/files#r504341236

Testing done:

```
Create PVC with reclaim policy retain:

kubectl create -f pvc-vc-sup-retain2.yaml
persistentvolumeclaim/pvc-vc-sup-retain2 created

# kubectl get pvc -A
NAMESPACE   NAME                          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE
default     pvc-vc-sup-retain2            Bound    pvc-55af66da-e9c8-42ba-93f2-01189618a55f   1Gi        RWO            vcp-sup-sc-retain   11s

# kubectl get cnsvspherevolumemigrations
NAME                                   AGE
e61d0264-0f44-45d5-8065-97a9055cdc5f   65s

Delete the PVC:

kubectl delete -f pvc-vc-sup-retain2.yaml
persistentvolumeclaim "pvc-vc-sup-retain2" deleted

Delete the PV and make sure DeleteVolumeInfo failed first time(induce temp failure for testing go routine):

# kubectl delete pv pvc-55af66da-e9c8-42ba-93f2-01189618a55f
persistentvolume "pvc-55af66da-e9c8-42ba-93f2-01189618a55f" deleted

Verified CR still exist:

root@k8s-master-965:~# kubectl get cnsvspherevolumemigrations -w
NAME                                   AGE
e61d0264-0f44-45d5-8065-97a9055cdc5f   53s

Wait for some time for the go routine to pick up the stale CR:

kubectl logs vsphere-csi-controller-88688d4dc-9pr2q -c vsphere-syncer -n kube-system | grep resourceToBeDeleted
2020-10-16T00:52:16.524Z	DEBUG	migration/migration.go:461	resourceToBeDeleted:  [e61d0264-0f44-45d5-8065-97a9055cdc5f]:	{"TraceId": "fbcc3161-7761-48b5-ab0f-8ff4f4c6f224"}

Verified that the go routine deleted the CR:
 # kubectl get cnsvspherevolumemigrations | grep "e61d0264-0f44-45d5-8065-97a9055cdc5f"
No resources found

Verified the same in logs:

2020-10-16T00:52:16.524Z	DEBUG	migration/migration.go:461	resourceToBeDeleted:  [e61d0264-0f44-45d5-8065-97a9055cdc5f]:	{"TraceId": "fbcc3161-7761-48b5-ab0f-8ff4f4c6f224"}
2020-10-16T00:52:16.542Z	DEBUG	migration/migration.go:156	received delete event for VolumeMigration CR!	{"TraceId": "4b594133-95f1-4d37-88dc-1d5d554edc93"}
2020-10-16T00:52:16.543Z	DEBUG	migration/migration.go:164	successfully deleted volumePath: "[vsanDatastore] 04b3715f-e838-12c9-f646-02009a32668b/4e958917ec844cfe87aecc73a52b3a47.vmdk", volumeID: "e61d0264-0f44-45d5-8065-97a9055cdc5f" mapping from cache	{"TraceId": "4b594133-95f1-4d37-88dc-1d5d554edc93"}
2020-10-16T00:52:16.546Z	INFO	migration/migration.go:468	Attempt to delete CR for e61d0264-0f44-45d5-8065-97a9055cdc5f	{"TraceId": "fbcc3161-7761-48b5-ab0f-8ff4f4c6f224"}
2020-10-16T00:52:16.546Z	INFO	migration/migration.go:470	Completed CnsVSphereVolumeMigration cleanup	{"TraceId": "fbcc3161-7761-48b5-ab0f-8ff4f4c6f224"}
**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
None
```

chethanv28 commented:
jtest block-vanilla

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 34 of 148 Specs in 6493.353 seconds
FAIL! -- 33 Passed | 1 Failed | 0 Pending | 114 Skipped
--- FAIL: TestE2E (6493.38s)
FAIL

Ginkgo ran 1 suite in 1h48m31.639682231s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/144/vsphere-csi-driver`

```

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+		return
+	}
+	go func() {
+		for {
```
instead of this kind of for look can use `for _ = range ticker.C {`

https://tutorialedge.net/golang/go-ticker-tutorial/

chethanv28 replied:
Done

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+			volumeMigrationResourceList := &migrationv1alpha1.CnsVSphereVolumeMigrationList{}
+			err = crdGroupClient.List(ctx, volumeMigrationResourceList)
+			if err != nil {
+				log.Warnf("failed to get CnsVSphereVolumeMigration list from the cluster. Err: %+v", err)
```
We should not  continue if we fail to obtain `volumeMigrationResourceList`

chethanv28 replied:
Why ? If we fail here, then the cleanup CRD routine will exit and we won't be able to delete/cleanup CRDs forever

divyenpatel replied:
but if you do not have volumeMigrationResourceList with you. How will you delete any CRs?

divyenpatel replied:
I mean end this instance of a job with `continue`

chethanv28 replied:
Okay, continue - sounds good to me

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+			queryAllResult, err := (*volumeMigrationInstance.volumeManager).QueryAllVolume(ctx, queryFilter, cnstypes.CnsQuerySelection{})
+			if err != nil {
+				log.Errorf("failed to queryAllVolume with err %+v", err)
+				return
```
we do not want to end the loop by returning from here. use `continue` instead.

chethanv28 replied:
Done

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+				}
+			}
+			log.Infof("Completed CnsVSphereVolumeMigration cleanup")
+			for i := 1; i <= volumeMigrationInstance.cnsConfig.Global.VolumeMigrationCRCleanupIntervalInHours; i++ {
```
Let's use `time.NewTicker` instead of adding a sleep this way.

chethanv28 replied:
Done

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+		log.Errorf("Failed to read config. Error: %+v", err)
+		return err
+	}
+	volumeMigrationInstance.cnsConfig = cfg
```
We have a volume manager holding a vCenter connection. I guess that will updated by metadata syncer during reloadconfig, as we have supplied pointer. Can you double check this.

divyenpatel commented on pkg/common/config/types.go:
```diff
@@ -43,6 +43,9 @@ type Config struct {
 		CnsRegisterVolumesCleanupIntervalInMin int `gcfg:"cnsregistervolumes-cleanup-intervalinmin"`
+		// VolumeMigrationCRCleanupIntervalInHours specifies the interval after which
+		// stale CnsVSphereVolumeMigration CRs will be cleaned up.
+		VolumeMigrationCRCleanupIntervalInHours int `gcfg:"volumemigration-cr-cleanup-intervalinhours"`
```
Let's set this in Minutes, and we will default to 60 minutes.

chethanv28 replied:
Done

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+	}()
+}
+
+// watcher watches on the vsphere.conf file mounted as secret
```
As discussed we do not need `watcher` and `reloadConfiguration` here in the volume migration service. Can you remove them? We already watch for config change in metadata syncer and controller where we are initializing volume migration service.

Volume Migration Service is initialized with a pointer for volume manager and config, so we are good.

```
GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config)
```

chethanv28 replied:
Removed the watcher and reload config method

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+
+//reloadConfiguration reloads configuration from the secret, and cnsOperator
+//with the latest configInfo
+func (volumeMigration *volumeMigration) reloadConfiguration(ctx context.Context) error {
```
we do not need reloadConfiguration as mentioned at https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/443/files#r506688900

chethanv28 replied:
+1

RaunakShah commented on example/vanilla-k8s-block-driver/vsphere.conf:
```diff
@@ -1,5 +1,6 @@
 [Global]
 cluster-id = "unique-kubernetes-cluster-id"
+volumemigration-cr-cleanup-intervalinhours = "24"
```
Why does this need to be configurable? Volume migration CR is an internal resource that only CSI driver knows about/uses.

divyenpatel replied:
volumemigration-cr-cleanup-intervalinhours => volumemigration-cr-cleanup-intervalinmin
24 => 120

chethanv28 replied:
Setting this to 60

RaunakShah commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+		log.Errorf("failed to get kubeconfig. err: %v", err)
+		return
+	}
+	crdGroupClient, err := k8s.NewClientForGroup(ctx, config, CRDGroupName)
```
Shouldn't this also be within the goroutine?

chethanv28 replied:
+1

RaunakShah commented on pkg/apis/migration/migration.go:
```diff
@@ -166,6 +171,7 @@ func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Man
 				stopCh := make(chan struct{})
 				informer.Informer().Run(stopCh)
 			}()
+			volumeMigrationInstance.cleanupStaleCRDInstances(ctx)
```
If i understand correctly, there's one instance of `volumeMigrationService` in both controller and syncer. In that case, we're going to have two cleanup goroutines running? Does it make sense to add this functionality to full sync, since its the same model?

divyenpatel replied:
controller and syncer run in two diff containers, so we have one instance per container.
it will be good if we run this clean up routine in only one container as @RaunakShah has suggested.

divyenpatel replied:
Can you supply control flag in GetVolumeMigrationService to decide if volumeMigrationInstance.cleanupStaleCRDInstances() routin needs to be called or not.

when called from controller init supply this flag as `false` and when called from syncer init supply this flag is `true`.

chethanv28 replied:
+1

chethanv28 replied:
```
SYNCER

# kubectl logs vsphere-csi-controller-587b968b4b-vbtzj -c vsphere-syncer -n kube-system | grep "Triggering"
2020-10-20T21:55:19.830Z	INFO	migration/migration.go:438	Triggering CnsVSphereVolumeMigration cleanup routine	{"TraceId": "683051b7-f7ed-46bb-bbc9-13aeb5704652"}
2020-10-20T21:56:19.830Z	INFO	migration/migration.go:438	Triggering CnsVSphereVolumeMigration cleanup routine	{"TraceId": "5316ce1e-2455-48df-9da1-d4daaebc078a"}
2020-10-20T21:57:19.829Z	INFO	migration/migration.go:438	Triggering CnsVSphereVolumeMigration cleanup routine	{"TraceId": "a52c0079-d650-49d6-a87e-ec3a2a937848"}
...
...
2020-10-20T22:52:19.830Z	INFO	migration/migration.go:438	Triggering CnsVSphereVolumeMigration cleanup routine	{"TraceId": "b03c3f1d-51ee-46e7-bd6b-976eccea7c96"}


CONTROLLER
root@k8s-master:~# kubectl logs vsphere-csi-controller-587b968b4b-vbtzj -c vsphere-csi-controller -n kube-system | grep "Triggering"
No results found
root@k8s-master:~#

RaunakShah commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+		log.Error("Failed to watch on config file for changes to VolumeMigrationCRCleanupIntervalInHours. Error: %+v", err)
+		return
+	}
+	go func() {
```
Can you verify that this goroutine ends when the container terminates?

chethanv28 replied:
Yes. It ends when the container is exited

RaunakShah commented:
/ok-to-test

divyenpatel commented on pkg/common/config/config.go:
```diff
@@ -62,6 +62,10 @@ const (
+	// DefaultVolumeMigrationCRCleanupIntervalInHours is the default time
+	// interval after which stale CnsVSphereVolumeMigration CRs will be cleaned up.
+	// Current default value is set to 24 hours
+	DefaultVolumeMigrationCRCleanupIntervalInHours = 24
```
DefaultVolumeMigrationCRCleanupIntervalInHours => DefaultVolumeMigrationCRCleanupIntervalInMinutes
24 => 120

chethanv28 replied:
Setting this to 60

divyenpatel commented on pkg/common/config/config.go:
```diff
@@ -329,6 +333,9 @@ func validateConfig(ctx context.Context, cfg *Config) error {
 		cfg.Global.CnsRegisterVolumesCleanupIntervalInMin = DefaultCnsRegisterVolumesCleanupIntervalInMin
 	}
+	if cfg.Global.VolumeMigrationCRCleanupIntervalInMin == 0 {
+		cfg.Global.VolumeMigrationCRCleanupIntervalInMin = DefaultVolumeMigrationCRCleanupIntervalInHours
```
change DefaultVolumeMigrationCRCleanupIntervalInHours to DefaultVolumeMigrationCRCleanupIntervalInMinutes

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +413,62 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
 }
+
+// cleanupStaleCRDInstances helps in cleaning up stale volume migration CRD instances
+func (volumeMigration *volumeMigration) cleanupStaleCRDInstances(ctx context.Context) {
```
make this entire `func` as routine, including the creation of the client to avoid timed client issues.

chethanv28 replied:
+1

chethanv28 commented:
Unit tests passing:

```
$ make unit-test
env -u VSPHERE_SERVER -u VSPHERE_DATACENTER -u VSPHERE_PASSWORD -u VSPHERE_USER -u VSPHERE_STORAGE_POLICY_NAME -u KUBECONFIG -u WCP_ENDPOINT -u WCP_PORT -u WCP_NAMESPACE -u TOKEN -u CERTIFICATE go test -v -count=1 ./pkg/common/config ./pkg/csi/service ./pkg/csi/service/common ./pkg/csi/service/common/commonco/k8sorchestrator ./pkg/csi/service/vanilla ./pkg/csi/service/wcp ./pkg/csi/service/wcpguest ./pkg/syncer ./pkg/syncer/admissionhandler
=== RUN   TestValidateConfigWithNoNetPermissionParams
{"level":"info","time":"2020-10-20T15:51:48.22597-07:00","caller":"config/config.go:296","msg":"No Net Permissions given in Config. Using default permissions."}
{"level":"info","time":"2020-10-20T15:51:48.226144-07:00","caller":"config/config.go:328","msg":"No feature states config information is provided in the Config. Using default config map name: internal-feature-states.csi.vsphere.vmware.com and namespace: kube-system"}
--- PASS: TestValidateConfigWithNoNetPermissionParams (0.00s)
=== RUN   TestValidateConfigWithMultipleNetPermissionParams
{"level":"info","time":"2020-10-20T15:51:48.226256-07:00","caller":"config/config.go:328","msg":"No feature states config information is provided in the Config. Using default config map name: internal-feature-states.csi.vsphere.vmware.com and namespace: kube-system"}
--- PASS: TestValidateConfigWithMultipleNetPermissionParams (0.00s)
=== RUN   TestValidateConfigWithInvalidPermissions
{"level":"error","time":"2020-10-20T15:51:48.226885-07:00","caller":"config/config.go:305","msg":"Invalid value WRITE_ONLY for Permissions under NetPermission Config A","stacktrace":"sigs.k8s.io/vsphere-csi-driver/pkg/common/config.validateConfig\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/common/config/config.go:305\nsigs.k8s.io/vsphere-csi-driver/pkg/common/config.TestValidateConfigWithInvalidPermissions\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/common/config/config_test.go:143\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:909"}
--- PASS: TestValidateConfigWithInvalidPermissions (0.00s)
=== RUN   TestValidateConfigWithInvalidClusterId
{"level":"error","time":"2020-10-20T15:51:48.22718-07:00","caller":"config/config.go:257","msg":"cluster id must not exceed 64 characters","stacktrace":"sigs.k8s.io/vsphere-csi-driver/pkg/common/config.validateConfig\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/common/config/config.go:257\nsigs.k8s.io/vsphere-csi-driver/pkg/common/config.TestValidateConfigWithInvalidClusterId\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/common/config/config_test.go:155\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:909"}
--- PASS: TestValidateConfigWithInvalidClusterId (0.00s)
PASS
ok  	sigs.k8s.io/vsphere-csi-driver/pkg/common/config	0.017s
=== RUN   TestGetDisk
=== RUN   TestGetDisk/#00
=== PAUSE TestGetDisk/#00
=== RUN   TestGetDisk/#01
=== PAUSE TestGetDisk/#01
=== CONT  TestGetDisk/#00
=== CONT  TestGetDisk/#01
--- PASS: TestGetDisk (0.00s)...
...
...
{"level":"info","time":"2020-10-20T15:52:08.731278-07:00","caller":"syncer/util.go:156","msg":"Metadata retrieved for all requested volumes"}
{"level":"info","time":"2020-10-20T15:52:08.7316-07:00","caller":"syncer/fullsync.go:413","msg":"FullSync: update is required for volume: \"b4439628-b0dd-42e5-8a17-5712211c17bd\""}
{"level":"info","time":"2020-10-20T15:52:08.731809-07:00","caller":"syncer/fullsync.go:228","msg":"FullSync: fullSyncDeleteVolumes could not find any volume which is not present in k8s and needs to be checked for volume deletion."}
{"level":"info","time":"2020-10-20T15:52:08.736084-07:00","caller":"volume/manager.go:513","msg":"UpdateVolumeMetadata: volumeID: \"b4439628-b0dd-42e5-8a17-5712211c17bd\", opId: \"\""}
{"level":"info","time":"2020-10-20T15:52:08.736108-07:00","caller":"volume/manager.go:531","msg":"UpdateVolumeMetadata: Volume metadata updated successfully. volumeID: \"b4439628-b0dd-42e5-8a17-5712211c17bd\", opId: \"\""}
{"level":"info","time":"2020-10-20T15:52:08.73612-07:00","caller":"syncer/fullsync.go:127","msg":"FullSync: end"}
{"level":"info","time":"2020-10-20T15:52:08.73882-07:00","caller":"volume/manager.go:449","msg":"DeleteVolume: volumeID: \"b4439628-b0dd-42e5-8a17-5712211c17bd\", opId: \"\""}
{"level":"info","time":"2020-10-20T15:52:08.738838-07:00","caller":"volume/manager.go:467","msg":"DeleteVolume: Volume deleted successfully. volumeID: \"b4439628-b0dd-42e5-8a17-5712211c17bd\", opId: \"\""}
--- PASS: TestSyncerWorkflows (12.17s)
    syncer_test.go:160: TestSyncerWorkflows: start
    syncer_test.go:278: TestMetadataSyncInformer start
    syncer_test.go:459: TestMetadataSyncInformer end
    syncer_test.go:607: TestFullSyncWorkflows start
    syncer_test.go:788: TestFullSyncWorkflows end
    syncer_test.go:248: TestSyncerWorkflows: end
=== RUN   TestValidMigratedAndLegacyVolume
--- PASS: TestValidMigratedAndLegacyVolume (0.00s)
=== RUN   TestGetSCNameFromPVC
--- PASS: TestGetSCNameFromPVC (0.00s)
    util_test.go:103: testGetSCNameFromPVC: start
    util_test.go:117: Verifying GetSCNameFromPVC for case where SC name is provided through only Spec.StorageClassName
    util_test.go:127: Verifying GetSCNameFromPVC for case where SC name is provided through only Metadata.Annotation
    util_test.go:138: Verifying GetSCNameFromPVC for case where SC name is provided through both Spec.StorageClassName and Metadata.Annotation
    util_test.go:149: Verifying GetSCNameFromPVC for case where SC name is provided through neither Spec.StorageClassName nor Metadata.Annotation
    util_test.go:158: testGetSCNameFromPVC: end
PASS
ok  	sigs.k8s.io/vsphere-csi-driver/pkg/syncer	12.219s
=== RUN   TestValidateStorageClassForAllowVolumeExpansion
{"level":"info","time":"2020-10-20T15:51:56.200533-07:00","caller":"admissionhandler/validatestorageclass.go:75","msg":"Validating StorageClass: \"sc\""}
{"level":"error","time":"2020-10-20T15:51:56.20077-07:00","caller":"admissionhandler/validatestorageclass.go:99","msg":"validation of StorageClass: \"sc\" Failed","stacktrace":"sigs.k8s.io/vsphere-csi-driver/pkg/syncer/admissionhandler.validateStorageClass\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/syncer/admissionhandler/validatestorageclass.go:99\nsigs.k8s.io/vsphere-csi-driver/pkg/syncer/admissionhandler.TestValidateStorageClassForAllowVolumeExpansion\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/syncer/admissionhandler/validatestorageclass_test.go:45\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:909"}
--- PASS: TestValidateStorageClassForAllowVolumeExpansion (0.00s)
    validatestorageclass_test.go:49: TestValidateStorageClassForAllowVolumeExpansion Passed
=== RUN   TestValidateStorageClassForMigrationParameter
{"level":"info","time":"2020-10-20T15:51:56.201075-07:00","caller":"admissionhandler/validatestorageclass.go:75","msg":"Validating StorageClass: \"sc\""}
{"level":"error","time":"2020-10-20T15:51:56.201098-07:00","caller":"admissionhandler/validatestorageclass.go:99","msg":"validation of StorageClass: \"sc\" Failed","stacktrace":"sigs.k8s.io/vsphere-csi-driver/pkg/syncer/admissionhandler.validateStorageClass\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/syncer/admissionhandler/validatestorageclass.go:99\nsigs.k8s.io/vsphere-csi-driver/pkg/syncer/admissionhandler.TestValidateStorageClassForMigrationParameter\n\t/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/pkg/syncer/admissionhandler/validatestorageclass_test.go:60\ntesting.tRunner\n\t/usr/local/go/src/testing/testing.go:909"}
--- PASS: TestValidateStorageClassForMigrationParameter (0.00s)
    validatestorageclass_test.go:64: TestValidateStorageClassForMigrationParameter Passed
=== RUN   TestValidateStorageClassForValidStorageClass
{"level":"info","time":"2020-10-20T15:51:56.201366-07:00","caller":"admissionhandler/validatestorageclass.go:75","msg":"Validating StorageClass: \"sc\""}
{"level":"info","time":"2020-10-20T15:51:56.201394-07:00","caller":"admissionhandler/validatestorageclass.go:97","msg":"Validation of StorageClass: \"sc\" Passed"}
--- PASS: TestValidateStorageClassForValidStorageClass (0.00s)
    validatestorageclass_test.go:79: TestValidateStorageClassForValidStorageClass Passed
PASS
ok  	sigs.k8s.io/vsphere-csi-driver/pkg/syncer/admissionhandler	0.112s
```

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +417,61 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+	log := logger.GetLogger(ctx)
+	ticker := time.NewTicker(time.Duration(volumeMigrationInstance.cnsConfig.Global.VolumeMigrationCRCleanupIntervalInMin) * time.Minute)
+	for range ticker.C {
+		config, err := k8s.GetKubeConfig(ctx)
```
Can we create a new context in this routine to avoid facing timed out context and track logs from each different cleanup cycle?

for example - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/syncer/metadatasyncer.go#L275-L276

chethanv28 replied:
Got it, thanks

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +417,61 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+			log.Warnf("failed to create new client for group %q . Err: %+v", CRDGroupName, err)
+			continue
+		}
+		ctx, log = logger.GetNewContextWithLogger()
```
this is for https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/443/files#r508892377

chethanv28 replied:
Done

divyenpatel commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +417,61 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+
+// cleanupStaleCRDInstances helps in cleaning up stale volume migration CRD instances
+func (volumeMigration *volumeMigration) cleanupStaleCRDInstances(ctx context.Context) {
+	log := logger.GetLogger(ctx)
```
you can remove this log and also let's not supply `ctx` to this routine method.

chethanv28 replied:
Makes sense

chethanv28 commented:
/test pull-vsphere-csi-driver-verify-staticcheck

chethanv28 commented:
/test pull-vsphere-csi-driver-verify-golangci-lint

chethanv28 commented:
/retest

chethanv28 commented:
Ran e2e test after addressing comments. Also, this PR is modifying the migration workflow for which e2e tests are not yet introduced. Hence running basic test to verify nothing is broken

```
$ make test-e2e
hack/run-e2e-test.sh
go: finding gopkg.in/tomb.v1 latest
go: finding golang.org/x/sys latest
Oct 20 16:29:49.370: INFO: The --provider flag is not set. Continuing as if --provider=skeleton had been used.
W1020 16:29:49.370046   57821 test_context.go:414] Unable to find in-cluster config, using default host : http://127.0.0.1:8080
I1020 16:29:49.370153   57821 test_context.go:423] Tolerating taints "" when considering if nodes are ready
Running Suite: CNS CSI Driver End-to-End Tests
==============================================
Random Seed: 1603236580 - Will randomize all specs
Will run 1 of 149 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
Data Persistence 
  [csi-block-vanilla] [csi-supervisor] [csi-guest] Data persistence vanilla test
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/data_persistence.go:107
[BeforeEach] Data Persistence
  /Users/chethanv/go/pkg/mod/k8s.io/kubernetes@v1.18.5/test/e2e/framework/framework.go:178
STEP: Creating a kubernetes client
Oct 20 16:29:49.385: INFO: >>> kubeConfig: /Users/chethanv/.kube/config
STEP: Building a namespace api object, basename e2e-data-persistence
Oct 20 16:29:49.526: INFO: Found PodSecurityPolicies; assuming PodSecurityPolicy is enabled.
Oct 20 16:29:49.591: INFO: Found ClusterRoles; assuming RBAC is enabled.
STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in e2e-data-persistence-8081
STEP: Waiting for a default service account to be provisioned in namespace
[BeforeEach] Data Persistence
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/data_persistence.go:72
[It] [csi-block-vanilla] [csi-supervisor] [csi-guest] Data persistence vanilla test
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/data_persistence.go:107
STEP: Creating Storage Class and PVC
STEP: CNS_TEST: Running for vanilla k8s setup
STEP: Creating StorageClass  with scParameters: map[] and allowedTopologies: [] and ReclaimPolicy:  and allowVolumeExpansion: false
STEP: Creating PVC using the Storage Class sc-j77jt with disk size  and labels: map[] accessMode: 
STEP: Waiting for claim pvc-lx6t6 to be in bound phase
Oct 20 16:29:50.107: INFO: Waiting up to 5m0s for PersistentVolumeClaims [pvc-lx6t6] to have phase Bound
Oct 20 16:29:50.123: INFO: PersistentVolumeClaim pvc-lx6t6 found but phase is Pending instead of Bound.
Oct 20 16:29:52.138: INFO: PersistentVolumeClaim pvc-lx6t6 found but phase is Pending instead of Bound.
Oct 20 16:29:54.155: INFO: PersistentVolumeClaim pvc-lx6t6 found but phase is Pending instead of Bound.
Oct 20 16:29:56.181: INFO: PersistentVolumeClaim pvc-lx6t6 found but phase is Pending instead of Bound.
Oct 20 16:29:58.209: INFO: PersistentVolumeClaim pvc-lx6t6 found and phase=Bound (8.101903002s)
STEP: Creating pod
STEP: Verify volume: 8542015a-88f5-4e19-8c7f-b8ee105f2136 is attached to the node: k8s-node1
STEP: VM UUID is: 4211afb6-f36a-3fe1-e147-6e90db3054b9 for node: k8s-node1
Oct 20 16:30:08.426: INFO: vmRef: VirtualMachine:vm-1024 for the VM uuid: 4211afb6-f36a-3fe1-e147-6e90db3054b9
Oct 20 16:30:08.486: INFO: Found FCDID "8542015a-88f5-4e19-8c7f-b8ee105f2136" attached to VM "k8s-node1"
Oct 20 16:30:08.486: INFO: Found the disk "8542015a-88f5-4e19-8c7f-b8ee105f2136" is attached to the VM with UUID: "4211afb6-f36a-3fe1-e147-6e90db3054b9"
STEP: Creating an empty file on the volume mounted on: pvc-tester-qvxgs
Oct 20 16:30:08.486: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/Users/chethanv/.kube/config exec --namespace=e2e-data-persistence-8081 pvc-tester-qvxgs -- /bin/sh -c touch /mnt/volume1/e2e-data-persistence-8081_file_A.txt'
Oct 20 16:30:09.136: INFO: stderr: ""
Oct 20 16:30:09.136: INFO: stdout: ""
STEP: Verify files exist on volume mounted on: pvc-tester-qvxgs
Oct 20 16:30:09.136: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/Users/chethanv/.kube/config exec --namespace=e2e-data-persistence-8081 pvc-tester-qvxgs -- /bin/ls /mnt/volume1/e2e-data-persistence-8081_file_A.txt'
Oct 20 16:30:09.512: INFO: stderr: ""
Oct 20 16:30:09.512: INFO: stdout: "/mnt/volume1/e2e-data-persistence-8081_file_A.txt\n"
STEP: Deleting the pod
Oct 20 16:30:09.512: INFO: Deleting pod "pvc-tester-qvxgs" in namespace "e2e-data-persistence-8081"
Oct 20 16:30:09.530: INFO: Wait up to 5m0s for pod "pvc-tester-qvxgs" to be fully deleted
STEP: Verify volume is detached from the node
STEP: VM UUID is: 4211afb6-f36a-3fe1-e147-6e90db3054b9 for node: k8s-node1
Oct 20 16:30:22.266: INFO: vmRef: VirtualMachine:vm-1024 for the VM uuid: 4211afb6-f36a-3fe1-e147-6e90db3054b9
Oct 20 16:30:22.306: INFO: Found FCDID "8542015a-88f5-4e19-8c7f-b8ee105f2136" attached to VM "k8s-node1"
Oct 20 16:30:22.306: INFO: Found the disk "8542015a-88f5-4e19-8c7f-b8ee105f2136" is attached to the VM with UUID: "4211afb6-f36a-3fe1-e147-6e90db3054b9"
Oct 20 16:30:22.306: INFO: Waiting for disk: "8542015a-88f5-4e19-8c7f-b8ee105f2136" to be detached from the node :"k8s-node1"
STEP: VM UUID is: 4211afb6-f36a-3fe1-e147-6e90db3054b9 for node: k8s-node1
Oct 20 16:30:23.701: INFO: vmRef: VirtualMachine:vm-1024 for the VM uuid: 4211afb6-f36a-3fe1-e147-6e90db3054b9
Oct 20 16:30:23.738: INFO: Found FCDID "8542015a-88f5-4e19-8c7f-b8ee105f2136" attached to VM "k8s-node1"
Oct 20 16:30:23.738: INFO: Found the disk "8542015a-88f5-4e19-8c7f-b8ee105f2136" is attached to the VM with UUID: "4211afb6-f36a-3fe1-e147-6e90db3054b9"
Oct 20 16:30:23.738: INFO: Waiting for disk: "8542015a-88f5-4e19-8c7f-b8ee105f2136" to be detached from the node :"k8s-node1"
STEP: VM UUID is: 4211afb6-f36a-3fe1-e147-6e90db3054b9 for node: k8s-node1
Oct 20 16:30:25.689: INFO: vmRef: VirtualMachine:vm-1024 for the VM uuid: 4211afb6-f36a-3fe1-e147-6e90db3054b9
Oct 20 16:30:25.724: INFO: failed to find FCDID "8542015a-88f5-4e19-8c7f-b8ee105f2136" attached to VM "k8s-node1"
Oct 20 16:30:25.724: INFO: Disk: 8542015a-88f5-4e19-8c7f-b8ee105f2136 successfully detached
STEP: Creating a new pod using the same volume
STEP: Verify volume: 8542015a-88f5-4e19-8c7f-b8ee105f2136 is attached to the node: k8s-node1
STEP: VM UUID is: 4211afb6-f36a-3fe1-e147-6e90db3054b9 for node: k8s-node1
Oct 20 16:30:37.908: INFO: vmRef: VirtualMachine:vm-1024 for the VM uuid: 4211afb6-f36a-3fe1-e147-6e90db3054b9
Oct 20 16:30:37.945: INFO: Found FCDID "8542015a-88f5-4e19-8c7f-b8ee105f2136" attached to VM "k8s-node1"
Oct 20 16:30:37.945: INFO: Found the disk "8542015a-88f5-4e19-8c7f-b8ee105f2136" is attached to the VM with UUID: "4211afb6-f36a-3fe1-e147-6e90db3054b9"
STEP: Creating a second empty file on the same volume mounted on: pvc-tester-fqq99
Oct 20 16:30:37.945: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/Users/chethanv/.kube/config exec --namespace=e2e-data-persistence-8081 pvc-tester-fqq99 -- /bin/sh -c touch /mnt/volume1/e2e-data-persistence-8081_file_B.txt'
Oct 20 16:30:38.380: INFO: stderr: ""
Oct 20 16:30:38.380: INFO: stdout: ""
STEP: Verify files exist on volume mounted on: pvc-tester-fqq99
Oct 20 16:30:38.380: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/Users/chethanv/.kube/config exec --namespace=e2e-data-persistence-8081 pvc-tester-fqq99 -- /bin/ls /mnt/volume1/e2e-data-persistence-8081_file_A.txt'
Oct 20 16:30:38.824: INFO: stderr: ""
Oct 20 16:30:38.824: INFO: stdout: "/mnt/volume1/e2e-data-persistence-8081_file_A.txt\n"
Oct 20 16:30:38.824: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/Users/chethanv/.kube/config exec --namespace=e2e-data-persistence-8081 pvc-tester-fqq99 -- /bin/ls /mnt/volume1/e2e-data-persistence-8081_file_B.txt'
Oct 20 16:30:39.249: INFO: stderr: ""
Oct 20 16:30:39.249: INFO: stdout: "/mnt/volume1/e2e-data-persistence-8081_file_B.txt\n"
STEP: Deleting the pod
Oct 20 16:30:39.249: INFO: Deleting pod "pvc-tester-fqq99" in namespace "e2e-data-persistence-8081"
Oct 20 16:30:39.268: INFO: Wait up to 5m0s for pod "pvc-tester-fqq99" to be fully deleted
STEP: Verify volume is detached from the node
STEP: VM UUID is: 4211afb6-f36a-3fe1-e147-6e90db3054b9 for node: k8s-node1
Oct 20 16:30:58.264: INFO: vmRef: VirtualMachine:vm-1024 for the VM uuid: 4211afb6-f36a-3fe1-e147-6e90db3054b9
Oct 20 16:30:58.299: INFO: failed to find FCDID "8542015a-88f5-4e19-8c7f-b8ee105f2136" attached to VM "k8s-node1"
Oct 20 16:30:58.299: INFO: Disk: 8542015a-88f5-4e19-8c7f-b8ee105f2136 successfully detached
Oct 20 16:30:58.299: INFO: Deleting PersistentVolumeClaim "pvc-lx6t6"
Oct 20 16:31:00.406: INFO: volume "8542015a-88f5-4e19-8c7f-b8ee105f2136" has successfully deleted
[AfterEach] Data Persistence
  /Users/chethanv/go/pkg/mod/k8s.io/kubernetes@v1.18.5/test/e2e/framework/framework.go:179
Oct 20 16:31:00.428: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-data-persistence-8081" for this suite.

• [SLOW TEST:71.097 seconds]
Data Persistence
/Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/data_persistence.go:60
  [csi-block-vanilla] [csi-supervisor] [csi-guest] Data persistence vanilla test
  /Users/chethanv/Downloads/csi-external/vsphere-csi-driver/tests/e2e/data_persistence.go:107
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 149 Specs in 71.098 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 148 Skipped
PASS

Ginkgo ran 1 suite in 1m20.335716325s
Test Suite Passed
```

divyenpatel reviewed (approved):
/approve

divyenpatel commented:
LGTM

RaunakShah commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +417,61 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+
+		log.Infof("Triggering CnsVSphereVolumeMigration cleanup routine")
+		volumeMigrationResourceList := &migrationv1alpha1.CnsVSphereVolumeMigrationList{}
+		err = crdGroupClient.List(ctx, volumeMigrationResourceList)
```
One question is instead of creating a new k8sclient, clientForGroup and Listing from API server, can we pass the informer created here - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/apis/migration/migration.go#L165
That informer contains a `Lister()` which i think can be reused..

chethanv28 replied:
the informer in (https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/apis/migration/migration.go#L165) 
is created inside a go routine(). Let's not reuse that.
Also, I think it's better to recreate k8sclient in cleanupCRDInstances go routine() so that we don't keep the client connection alive through out

RaunakShah commented on pkg/apis/migration/migration.go:
```diff
@@ -407,3 +417,61 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
+		for _, volumeMigrationResource := range volumeMigrationResourceList.Items {
+			if _, existsInCNSVolumesMap := cnsVolumesMap[volumeMigrationResource.Name]; !existsInCNSVolumesMap {
+				log.Debugf("Volume with id %s is not found in CNS", volumeMigrationResource.Name)
+				resourcesToBeDeleted = append(resourcesToBeDeleted, volumeMigrationResource.Name)
```
This is probably a nit but why not just delete the instance here?

chethanv28 replied:
Actually it makes sense to delete right there. Thanks

RaunakShah commented on example/vanilla-k8s-block-driver/vsphere.conf:
```diff
@@ -1,5 +1,6 @@
 [Global]
 cluster-id = "unique-kubernetes-cluster-id"
+volumemigration-cr-cleanup-intervalinmin = "120"
```
I think i had a comment here about the need for having this as configurable? This is an internal CRD that's not exposed to customers, so why do we want to expose the cleanup interval to them?

chethanv28 replied:
We have started moving the timeouts/intervals (internal/external ) to a conf file so that it can be configurable.
I think it's okay to these vars as configurable
For ex: https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/common/config/types.go#L45
Also, refer: https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/426#discussion_r505736425

divyenpatel replied:
this is not a mandatory config, basically, when things are not working as expected, we need some sort of way to tweak these internal configurations. 

we can also supply this from the optional command-line argument if we do not want to put this in the config.

divyenpatel replied:
@chethanv28 when user changes this interval, do change the frequency of the ticker in the routin?

chethanv28 replied:
Yes

RaunakShah commented on pkg/common/config/config.go:
```diff
@@ -62,6 +62,10 @@ const (
 	DefaultCnsRegisterVolumesCleanupIntervalInMin = 720
+	// DefaultVolumeMigrationCRCleanupIntervalInMin is the default time
+	// interval after which stale CnsVSphereVolumeMigration CRs will be cleaned up.
+	// Current default value is set to 24 hours
```
Nit: 2 hours?

chethanv28 replied:
Good catch! Thanks

divyenpatel commented:
/lgtm

divyenpatel commented:
/hold

divyenpatel commented:
/hold cancel

divyenpatel commented:
/lgtm
